src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs

```diff
@@ -320,52 +296,29 @@ void addArg(RefKind refKind, BoundExpression expression)
-                        WellKnownMember.System_Index__ctor,
-                        ImmutableArray.Create<BoundExpression>(
-                            _factory.Literal(Math.Abs(index)),
-                            _factory.Literal(fromEnd)));
-                }
+                    // LocalRewriter.MakePatternIndexOffsetExpression understands this format
+                    if (index < 0)
+                    {
+                        MethodSymbol symbolOpt = _factory.WellKnownMember(WellKnownMember.System_Index__ctor) as MethodSymbol;
+                        return new BoundFromEndIndexExpression(_factory.Syntax, _factory.Literal(-index),
+                            methodOpt: symbolOpt, _factory.WellKnownType(WellKnownType.System_Index));
+                    }
 
-                BoundExpression makeExplicitRangeArgument(BoundDagSliceEvaluation e)
-                {
-                    return _factory.New(
-                        WellKnownMember.System_Range__ctor,
-                        ImmutableArray.Create(
-                            makeExplicitIndexArgument(e.StartIndex, fromEnd: false),
-                            makeExplicitIndexArgument(e.EndIndex, fromEnd: true)));
+                    return _factory.Convert(_factory.WellKnownType(WellKnownType.System_Index), _factory.Literal(index));
                 }
 
-                BoundExpression makeIndexerAccess(SyntaxNode syntax, BoundIndexerAccess indexerAccess, BoundExpression indexerArg)
+                BoundExpression makeUnloweredRangeArgument(BoundDagSliceEvaluation e)
                 {
-                    ImmutableArray<BoundExpression> arguments = indexerAccess.Arguments;
-                    Debug.Assert(!arguments.IsDefaultOrEmpty);
-
-                    if (arguments[0] is BoundConversion conv)
-                        indexerArg = conv.UpdateOperand(indexerArg);
-
-                    return _localRewriter.MakeIndexerAccess(
-                        syntax,
-                        input,
-                        indexerAccess.Indexer,
-                        arguments.SetItem(0, indexerArg),
-                        indexerAccess.ArgumentNamesOpt,
-                        indexerAccess.ArgumentRefKindsOpt,
-                        indexerAccess.Expanded,
-                        indexerAccess.ArgsToParamsOpt,
-                        indexerAccess.DefaultArguments,
-                        indexerAccess.Type,
-                        oldNodeOpt: null,
-                        isLeftOfAssignment: false);
+                    // LocalRewriter.VisitRangeImplicitIndexerAccess understands this format
+                    MethodSymbol symbolOpt1 = _factory.WellKnownMember(WellKnownMember.System_Index__ctor) as MethodSymbol;
+                    var end = new BoundFromEndIndexExpression(_factory.Syntax, _factory.Literal(-e.EndIndex),
+                        methodOpt: symbolOpt1, _factory.WellKnownType(WellKnownType.System_Index));
```

Comment:
Can we simply delegate this work to ```makeUnloweredIndexArgument```?

Reply:
I'd tried that but ran into problems. 
`makeUnloweredIndexArgument` deals with both indexes from start and from end, and the indexes from end are represented with negative values (-1, -2, etc, but not 0).
But for ranges we also need to represent zero-from-end.